File: SOLID/TS/OCP/DocumentPipeline.ts
import { Document } from "./Document";
import { DocumentStep } from "./DocumentStep";

export class DocumentPipeline {
    constructor(private steps: DocumentStep[]) { }

    process(document: Document): void {
        for (const step of this.steps) {
            step.execute(document);
        }
    }
}

File: SOLID/Java/OCP/DocumentPipeline.java
package OCP;

import java.util.List;

public class DocumentPipeline {

    private final List<DocumentStep> steps;

    public DocumentPipeline(List<DocumentStep> steps) {
        this.steps = steps;
    }

    public void process(Document document) {
        for (DocumentStep step : steps) {
            step.execute(document);
        }
    }
}
